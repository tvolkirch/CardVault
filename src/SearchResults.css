#search-results
{
    width: 80%;
    margin-top: 1rem;
}

#matchedCardList
{
    list-style-type: none;
}

#matchedCardList li
{
    margin-bottom: 0.5rem;
}

.card-overlay
{
    position: absolute;
    display: none;
    top: 0;
    left: 0;
    height: 100vh;
    width: 100vw;
    background: rgba(255, 255, 255, 0.95);
    color: #333;
    margin: 0;
    padding: 1rem;
    z-index: 3;
}

.matched-card-name
{
    display: inline-block;
}

.matched-card-name:hover
{
    cursor: pointer;
    text-decoration: underline;
}
