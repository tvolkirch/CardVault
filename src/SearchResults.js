import React from "react";

import Card from "./Card";

import { getPressedKeyStatus } from "./lib/EventManager";

import "./SearchResults.css";

function SearchResults(props)
{
    var openCardOverlay = function(index)
    {
        var id = "cardOverlay" + index;
        var cardOverlay = document.getElementById(id);
        var searchOverlay = document.getElementById("search-overlay");
        var buttonArray = cardOverlay.getElementsByClassName("close-button");
        const currentCard = document.getElementById("cardOverlay" + index);
        var cardArray = currentCard.getElementsByClassName("card-content card-text");

        searchOverlay.className = "invisible";
        cardOverlay.style.display = "block";

        if (buttonArray && buttonArray[0])
        {
            buttonArray[0].focus();
            buttonArray[0].addEventListener("keydown", handleCloseButton);
        }

        if ( cardArray && cardArray[0] )
        {
            cardArray[0].addEventListener("keydown", handleCardTabbing);
        }
    }

    var closeCardOverlay = function(index)
    {
        var id = "cardOverlay" + index;
        var cardOverlay = document.getElementById(id);
        var searchOverlay = document.getElementById("search-overlay");
        var buttonArray = cardOverlay.getElementsByClassName("close-button");
        var searchText = document.getElementById("searchText");
        const currentCard = document.getElementById("cardOverlay" + index);
        var cardArray = currentCard.getElementsByClassName("card-content card-text");

        searchOverlay.className = "";
        cardOverlay.style.display = "none";

        if (buttonArray && buttonArray[0])
        {
            buttonArray[0].removeEventListener("keydown", handleCloseButton);
        }

        if ( cardArray && cardArray[0] )
        {
            cardArray[0].removeEventListener("keydown", handleCardTabbing);
        }

        searchText.focus();
    }

    function handleCloseButton(eve)
    {
        const originatorElement = eve.target;
        const index = originatorElement.getAttribute("data");
        const originatorClass = originatorElement.getAttribute("class");
        const currentCard = document.getElementById("cardOverlay" + index);
        var cardArray = currentCard.getElementsByClassName("card-content card-text");
        const keyChecker = getPressedKeyStatus(eve);

        if ( originatorClass.indexOf("close-button") > -1 )
        {
            if ( keyChecker.isEnter || keyChecker.isSpace )
            {
                closeCardOverlay(index);
            }

            if ( cardArray && cardArray[0] && keyChecker.isTab && eve.shiftKey )
            {
                eve.preventDefault();
                cardArray[0].focus();
            }
        }
    }

    function handleCardTabbing(eve)
    {
        const originatorElement = eve.target;
        const index = originatorElement.getAttribute("data");
        const currentCard = document.getElementById("cardOverlay" + index);
        var closeButtonArr = currentCard.getElementsByClassName("close-button");
        const keyChecker = getPressedKeyStatus(eve);

        if (closeButtonArr && closeButtonArr[0] )
        {
            if (keyChecker.isTab && !eve.shiftKey)
            {
                eve.preventDefault();
                closeButtonArr[0].focus();
            }
        }
    }

    var cards = "";

    if (props.cards && props.cards.length > 0)
    {
       cards = props.cards.map((card, index) =>
           <li key={index}>
               <div className="matched-card-name" id={"matchedCardName" + index} data={index} tabIndex="0"
                    onClick={(eve) => openCardOverlay(index)}>{card.name}</div>
               <div className="card-overlay" id={"cardOverlay" + index}>
                    <div className="close-button" tabIndex="0" data={index}
                         onClick={(eve) => closeCardOverlay(index)}>
                        <span className="close-icon" aria-label="close button" role="button">&#8855;</span>
                    </div>
                    <Card
                     index={index}
                     url={card.imageUrl}
                     name={card.name}
                     type={card.type}
                     set={card.set.name}
                     rules={card.text}/>
               </div>
           </li>
       );
    }

    var message = "Click on any card name to display the card.";

    if (props.count === 0)
    {
        message = "No card names were found that matched the search text.";
    }
    else if (props.count >= 10)
    {
        message = "The search text matched " + props.count +
           " card names. No more than 10 matches will be returned. " + message;
    }
    else if (props.count === 1)
    {
        message = "The search text matched " + props.count + " card name.  Click on the card name to display the card.";
    }
    else
    {
        message = "The search text matched " + props.count + " card names. " + message;
    }

    return (
        <div>
            <div id="search-results" tabIndex="0">{message}</div>
            <ul id="matchedCardList">{cards}</ul>
        </div>
    );
}

export default SearchResults;