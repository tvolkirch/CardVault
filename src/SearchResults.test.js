import React from "react";
import renderer from "react-test-renderer";

import SearchResults from "./SearchResults";

it("Consistently renders search results", () => {

   const count = "1";
   var cards = [];
   var card = {};

   card.imageUrl = "#";
   card.name = "Card Name";
   card.type = "Card Type";
   card.set = {};
   card.set.name = "Set Name";
   card.text = "Card Text";

   cards[0] = card;

   const searchResults = renderer
      .create(<SearchResults
          count={count}
          cards={cards} />)
      .toJSON();

   expect(searchResults).toMatchSnapshot();
});
