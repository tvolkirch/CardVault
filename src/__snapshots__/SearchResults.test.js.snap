// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consistently renders search results 1`] = `
<div>
  <div
    id="search-results"
    tabIndex="0"
  >
    The search text matched 1 card names. Click on any card name to display the card.
  </div>
  <ul
    id="matchedCardList"
  >
    <li>
      <div
        className="matched-card-name"
        data={0}
        id="matchedCardName0"
        onClick={[Function]}
        tabIndex="0"
      >
        Card Name
      </div>
      <div
        className="card-overlay"
        id="cardOverlay0"
      >
        <div
          className="close-button"
          data={0}
          onClick={[Function]}
          tabIndex="0"
        >
          <span
            aria-label="close button"
            className="close-icon"
            role="button"
          >
            âŠ—
          </span>
        </div>
        <div
          className="flex-card-wrapper"
        >
          <div
            className="card-content"
          >
            <img
              alt="Elder Scrolls Card: Card Name"
              className="card-image"
              src="#"
              tabIndex="0"
            />
          </div>
          <div
            className="card-content card-text"
            data={0}
            tabIndex="0"
          >
            <div
              className="card-name"
            >
              Card Name
            </div>
            <div>
              <div
                className="flex-card-row"
              >
                <span
                  className="left-column"
                >
                  Type:
                </span>
                <span
                  className="right-column"
                >
                  Card Type
                </span>
              </div>
              <div
                className="flex-card-row"
              >
                <span
                  className="left-column"
                >
                  Set:
                </span>
                <span
                  className="right-column"
                >
                  Set Name
                </span>
              </div>
            </div>
            <div
              className="card-rules"
            >
              Card Text
            </div>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
